Add some capabilities for the builtin functions (#457)
unicode-org/message-format-wg: exploration/0001-builtin-registry-capabilities.md

# Capabilities for built-ins

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@ryzokuken</dd>
		<dt>First proposed</dt>
		<dd>2023-08-22</dd>
		<dt>Pull Request</dt>

> 		<dd>#457</dd>

Reviewer: ```suggestion
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/457">#457</a></dd>
```

	</dl>
</details>

## Objective

> Document the capabilities that the set of built-in functions would expose to the user in order to ensure it can be designed in the most minimalistic way that would still cater to the most pressing requirements, partly inspired by ICU4X anbd JavaScript's Intl API and backwards-compatible with MF1 in ICU4J.

Reviewer: ```suggestion
Document the capabilities that the set of built-in functions would expose to the user in order to ensure it can be designed in the most minimalistic way that would still cater to the most pressing requirements, partly inspired by ICU4X and JavaScript's Intl API and backwards-compatible with MF1 in ICU4J.
```

This document aims to list down the basic capabilities for developers that such a set of standard built-in functions MUST provide.

## Use-Cases

From formatting date and time values to performing selection based on plural rules, the use cases for a minimal set of common tools standardized across implementations are virtually everywhere. While a number of applications might require more complex functions with custom logic, a decently designed set of built-in functions could significantly lower the barrier of entry as well as the need to write many custom functions.

## Requirements

TBD

## Constraints

TBD

## Proposed Design

All conforming implementations of this specification MUST provide a set of <dfn>built-in functions</dfn>, which allows the developers to perform the following operations:

* Formatting the following types of data in a locale-sensitive manner:
    * Dates and times
    * Numbers
    * Lists
* Selecting between pattern strings in a locale-sensitive manner based on:
	* String equality
    * Plural rules

>     * Ordinality

Reviewer: Isn't ordinality a subset of plural rules? I know that MF1 has separately `plural` and `selectordinal`, but wouldn't it make sense for our plural selector to include something like `type=ordinal` as an option instaed?

Author: In my basic understanding of linguistics, there's not a whole lot of overlap between the two concepts. We could obviously decide to design the API either way, I have no strong opinions but we should err on the side of not confusing the user between two otherwise distinct linguistic concepts.

Reviewer 2: @eemeli Yes, you're correct and we might cover it using an option. But it is good to ensure we remember it on the checklist of things we need to build in.

That said, there are a number of common i18n operations that are quite universal and there's a lot of value in standardizing the usage of these common operations across the ecosystem.
This document aims to list down the basic capabilities for developers that such a set of standard built-in functions MUST provide.

## Use-Cases

From formatting date and time values to performing selection based on plural rules, the use cases for a minimal set of common tools standardized across implementations are virtually everywhere. While a number of applications might require more complex functions with custom logic, a decently designed set of built-in functions could significantly lower the barrier of entry as well as the need to write many custom functions.

## Requirements

TBD

## Constraints

TBD

## Proposed Design

All conforming implementations of this specification MUST provide a set of <dfn>built-in functions</dfn>, which allows the developers to perform the following operations:

- Formatting the following types of data in a locale-sensitive manner:
  - Dates and times
  - Numbers
  - Lists
- Selecting between pattern strings in a locale-sensitive manner based on:

>   - String equality
>     - Plural rules
>     - Ordinality

Reviewer: ```suggestion
  - String equality
  - Plural rules
  - Ordinality
```

functions.
This includes ensuring sufficient compatibility with the various functions available in ICU4X, JavaScript `Intl` as well
as MessageFormat v1.

## Background

Users have the ability to specify and call arbitrary custom functions in their messages thanks to the function registry.
That said, there are a number of common i18n operations that are quite universal and there's a lot of value in
standardizing the usage of these common operations across the ecosystem.
This document aims to list down the basic capabilities that implementations need to be provide in a _built-in registry_.

## Use-Cases

From formatting date and time values to performing selection based on plural rules, the use cases for a minimal set of
common tools standardized across implementations are virtually everywhere.
While a number of applications might require more complex functions with custom logic, a well-provisioned default
registry could significantly lower the barrier of entry as well as the need to write and maintain many custom functions.

## Requirements

Implementations are required to implement and expose the entire _built-in registry_ that would provide the capabilities

> in this document.

Reviewer 2: The requirements section should be requirements on this design, rather than what this design requires. A comprehensive set of requirements could be rather extensive, but it's still a good idea to spend the effort. One way to do that is to populate this section as you write the design.

## Constraints

> Like many standardized APIs out there, the biggest constraint on the resulting _built-in registry_ would be that until
> we decide to make another major update to MessageFormat, we'd not be able to make breaking changes to the function
> signatures since that would invalidate the promise of a consistent API across implementations and versions.

Reviewer 2: In some ways, this is too weak. It also belongs in requirements.

```suggestion
The names of functions and associated options defined by the registry
will be embedded into messages across the MF2 ecosystem.
These function signatures **_MUST_** be stable and cannot be changed
later.
```

Another constraint is that since MessageFormat would be used alongside other internationalization features in many

> programming environments, the final interface should be somewhat consistent to reduce developer confusion.

Reviewer 2: Not sure what the constraint is exactly?

## Proposed Design

The **_<dfn>built-in registry</dfn>_** is the set of functions, together with their options, that are present in every
implementation.
Implementations MUST provide each of the functions defined in the _built-in registry_, including all of the options.

> The ouptut of each function is implementation-defined.

Reviewer 2: ```suggestion
The output of each function is implementation-defined.
```

(typo)

> They MAY include additional options, although implementers are cautioned to avoid options which might conflict with
> future standardization.

Reviewer 2: tweak

```suggestion
Implementations MAY include additional options for built-in functions,
although implementers are cautioned to avoid adding options
that might conflict with future standardization.
```

## Requirements

Implementations are required to implement and expose the entire _built-in registry_ that would provide the capabilities in this document.

## Constraints

Like many standardized APIs out there, the biggest constraint on the resulting set of _built-in functions_ would be that until we decide to make another major update to MessageFormat, we'd not be able to make breaking changes to the function signatures since that would invalidate the promise of a consistent API across implementations and versions.

Another constraint is that since MessageFormat would be used alongside other internationalization features in many programming environments, the final interface should be somewhat consistent to reduce developer confusion.

## Proposed Design

The **_<dfn>built-in registry</dfn>_** is the set of _functions_, together with their options, that are present in every implementation.
Implementations MUST provide each of the built-in functions defined in the _built-in registry_, including all of the options.

> They MAY include additional options, although implementers are cautioned to avoid options which might conflict with future standardization.

Reviewer: This is fine, but could include a clarification (potentially as a later change) that the exact output produced for any of these functions or options is left up to each implementation.

So for example a conformant implementation could "support" the `notation` option for `:number` by not throwing an error if it's used, but producing the same output no matter what the value the option has.

## Alternatives Considered

> The alternative would be to not include any _built-in functions_, essentially outsourcing the task of curating and maintaining commonly used functions to the ecosystem.

Reviewer: Could also list as an alternative the preceding form of this PR, which defines but does not require the interfaces.

## Alternatives Considered

The alternative would be to not include any _built-in functions_, essentially outsourcing the task of curating and maintaining commonly used functions to the ecosystem.

> Another alternative would be to specify the signatures of the functions in a default registry for interoperability between implementations but not require implementations to implement all of them.

Reviewer 2: This could use more work.

```suggestion
The working group has considered a number of ways to address core functionality. These include:

**No default registry** Each implementation would be free to define
its own set of functions and options for each.

- (+) Each implementation would be able to define functions
  and options according to existing I18N APIs
  which would be familiar to users in that language or runtime
- (-) Messages would not be portable between implementations
- (-) Users would have to learn the variations between implementations
- (-) Tooling would have to be adjusted for each variation

**Informative registry** The default registry could be made informative
with implementations allowed to pick-and-choose (or ignore) entries.
This would provide better interoperability for implementations 
claiming to implement the default registry, while allowing platform-specific
variation.

**Non-extensible registry entries.** Like this design, but not allowing
implementation specific options. 
 
```
